activerecord/lib/active_record/attribute_methods/query.rb

```diff
@@ -16,17 +16,10 @@ def query_attribute(attr_name)
         when true        then true
         when false, nil  then false
         else
-          column = self.class.columns_hash[attr_name]
-          if column.nil?
-            if Numeric === value || value !~ /[^0-9]/
-              !value.to_i.zero?
-            else
-              return false if ActiveModel::Type::Boolean::FALSE_VALUES.include?(value)
-              !value.blank?
-            end
-          elsif value.respond_to?(:zero?)
-            !value.zero?
+          if Numeric === value || (String === value && value !~ /[^0-9]/)
+            !value.to_i.zero?
           else
+            return false if ActiveModel::Type::Boolean::FALSE_VALUES.include?(value)
```

Comment:
Can you add an assertion like here? d3964704ae2f5456dee2262fc34c8b910d82c01a would break the existing behavior.

```diff
diff --git a/activerecord/test/cases/attribute_methods_test.rb b/activerecord/test/cases/attribute_methods_test.rb
index 82ccb3605c..ca8c21d77f 100644
--- a/activerecord/test/cases/attribute_methods_test.rb
+++ b/activerecord/test/cases/attribute_methods_test.rb
@@ -433,6 +433,10 @@ def topic.title() "b" end
     end

     assert_equal true, Topic.new(author_name: "Name").author_name?
+
+    ActiveModel::Type::Boolean::FALSE_VALUES.each do |value|
+      assert_equal true, Topic.new(author_name: value).author_name?
+    end
   end

   test "number attribute predicate" do
```

```
% ARCONN=sqlite3 bundle exec ruby -w -Itest test/cases/attribute_methods_test.rb -n test_string_attribute_predicate
Using sqlite3
Run options: -n test_string_attribute_predicate --seed 26527

# Running:

F

Failure:
AttributeMethodsTest#test_string_attribute_predicate [test/cases/attribute_methods_test.rb:438]:
Expected: true
  Actual: false


rails test test/cases/attribute_methods_test.rb:430



Finished in 0.068753s, 14.5448 runs/s, 72.7241 assertions/s.
1 runs, 5 assertions, 1 failures, 0 errors, 0 skips
```

Reply:
@kamipo I apologize for late reply and fixing.
I checked query attribute method behavior by below code. There are no error, so I add new assertion here.
I would like to you to check this code is valid.

```
begin
  require 'bundler/inline'
rescue LoadError
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise
end

gemfile(true) do
  source 'https://rubygems.org'

  # Activate the gem you are reporting the issue against.
  gem 'rails', github: 'rails/rails'
  gem 'sqlite3'
  gem 'pry'
end

require 'active_record'
require 'pry'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database:  ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true
end

class Post < ActiveRecord::Base
  attribute :user_defined_string, :string
end

class BugTest < Minitest::Test
  def test_without_transaction
    binding.pry
    post = Post.new

    assert_equal post.user_defined_string?, false

    ActiveModel::Type::Boolean::FALSE_VALUES.each do |v|
      post = Post.new(user_defined_string: v)
      assert_equal post.user_defined_string?, false
    end
  end
end
```

Reply:
No, this PR still break the existing behavior 812e59bfa3a3462452b2b5e47e2937cd994f2aa9.
Please consider about the my comment https://github.com/rails/rails/pull/35320#discussion_r257924552.